src/fp6.rs

```diff
@@ -270,12 +270,48 @@ impl Fp6 {
         self.c0.is_zero() & self.c1.is_zero() & self.c2.is_zero()
     }
 
-    /// Returns `c = self * b`.
-    ///
-    /// Implements the full-tower interleaving strategy from
-    /// [ePrint 2022-376](https://eprint.iacr.org/2022/367).
     #[inline]
     fn mul_interleaved(&self, b: &Self) -> Self {
+        cfg_if::cfg_if! {
+                if #[cfg(target_os = "zkvm")] {
+                    // Implements Algorithm 13 from https://eprint.iacr.org/2010/354.pdf
+                    let mut t0 = self.c0;
+                    t0.mul_inp(&b.c0);
+                    let mut t1 = self.c1;
+                    t1.mul_inp(&b.c1);
+                    let mut t2 = self.c2;
+                    t2.mul_inp(&b.c2);
+                    let mut c0 = self.c1;
+                    c0.add_inp(&self.c2);
+                    let mut tmp = b.c1;
+                    tmp.add_inp(&b.c2);
+                    c0.mul_inp(&tmp);
+                    tmp = t2;
+                    tmp.add_inp(&t1);
+                    c0.sub_inp(&tmp);
+                    c0 = c0.mul_by_nonresidue();
+                    c0.add_inp(&t0);
+                    let mut c1 = self.c0;
+                    c1.add_inp(&self.c1);
+                    tmp = b.c0;
+                    tmp.add_inp(&b.c1);
+                    c1.mul_inp(&tmp);
+                    tmp = t0;
+                    tmp.add_inp(&t1);
+                    c1.sub_inp(&tmp);
+                    tmp = t2.mul_by_nonresidue();
+                    c1.add_inp(&tmp);
+                    tmp = self.c0;
+                    tmp.add_inp(&self.c2);
+                    let mut c2 = b.c0;
+                    c2.add_inp(&b.c2);
+                    c2.mul_inp(&tmp);
+                    tmp = t0;
+                    tmp.add_inp(&t2);
+                    c2.sub_inp(&tmp);
+                    c2.add_inp(&t1);
+                    Fp6 { c0, c1, c2 }
+                } else {
```

Comment:
I'd add a more detailed comment for what's going on here. Especially given that many people will be using this library it's important to be clear about how you got this (e.g. addchain).

Reply:
![Screenshot 2024-08-19 at 5 09 00 PM](https://github.com/user-attachments/assets/aa78aa07-3310-47d7-a8c8-6b92e69b07b4)

Oh I just directly implemented this algorithm verbatim